phoenix-core/src/it/java/org/apache/phoenix/end2end/index/GlobalIndexCheckerIT.java

```diff
@@ -317,50 +358,94 @@ public void testOnePhaseOverwrite() throws Exception {
+                IndexToolIT.runIndexTool(true, false, null, dataTableName, indexName + "2");
+            }
+            // Configure IndexRegionObserver to fail the last two write phase (i.e., the data table update and post index update phase)
+            // and check that this does not impact the correctness
+            IndexRegionObserver.setFailDataTableUpdatesForTesting(true);
+            IndexRegionObserver.setFailPostIndexUpdatesForTesting(true);
+            conn.createStatement().execute("upsert into " + dataTableName + " (id, val2) values ('a', 'abcc')");
+            commitWithException(conn);
+            IndexRegionObserver.setFailDataTableUpdatesForTesting(false);
+            IndexRegionObserver.setFailPostIndexUpdatesForTesting(false);
+            conn.createStatement().execute("upsert into " + dataTableName + " (id, val3) values ('a', 'abcdd')");
+            conn.commit();
+            String selectSql = "SELECT val2, val3 from " + dataTableName + " WHERE val1  = 'ab'";
+            // Verify that we will read from the first index table
+            assertExplainPlan(conn, selectSql, dataTableName, indexName + "1");
+            ResultSet rs = conn.createStatement().executeQuery(selectSql);
+            assertTrue(rs.next());
+            assertEquals("abc", rs.getString(1));
+            assertEquals("abcdd", rs.getString(2));
+            assertFalse(rs.next());
+            selectSql = "SELECT val2, val3 from " + dataTableName + " WHERE val2  = 'abc'";
+            // Verify that we will read from the second index table
+            assertExplainPlan(conn, selectSql, dataTableName, indexName + "2");
+            rs = conn.createStatement().executeQuery(selectSql);
+            assertTrue(rs.next());
+            assertEquals("abc", rs.getString(1));
+            assertEquals("abcdd", rs.getString(2));
+            assertFalse(rs.next());
+            // Add rows and check everything is still okay
+            verifyTableHealth(conn, dataTableName, indexName);
         }
-        // Configure Indexer to skip the last two write phase (i.e., the data table update and post index update phase)
-        // and check that this does not impact the correctness
-        IndexRegionObserver.setSkipDataTableUpdatesForTesting(true);
-        IndexRegionObserver.setSkipPostIndexUpdatesForTesting(true);
-        conn.createStatement().execute("upsert into " + dataTableName + " (id, val2) values ('a', 'abcc')");
-        conn.commit();
-        IndexRegionObserver.setSkipDataTableUpdatesForTesting(false);
-        IndexRegionObserver.setSkipPostIndexUpdatesForTesting(false);
-        conn.createStatement().execute("upsert into " + dataTableName + " (id, val3) values ('a', 'abcdd')");
+    }
+
+    static private void commitWithException(Connection conn) {
+        try {
+            conn.commit();
+            IndexRegionObserver.setFailPreIndexUpdatesForTesting(false);
```

Comment:
nit: could be useful to have these three lines setting to false in an After method to guarantee that they're always reset between tests.

Reply:
It sounds good. I will do that.

Reply:
Done.